### PR TITLE
Add DAG-based kernel cost model for layout optimization

### LINKED ISSUE #2316: layout-optimization: Add a proper cost model for the cost of a kernel
Currently we only have 1 meaningful kernel per linalg operation, but the layout-optimization pass has the ability to select a different kernel based on its cost and the cost of associated layout changes of operands and results.

We should upgrade the pass to have a proper cost model, in preparation for other operations which have many kernels (e.g., ct-ct matrix multiplication). These kernels will largely differ based on the data size being packed into them, as small matrices can be packed into an entire ciphertext while large matrices need to be split across many.

### PR DESCRIPTION
**Pull request by bon-cdp**
## Summary - updated from using implementMatvec to [implementHaleviShoup](https://github.com/google/heir/blob/7aa0fe8285e0df05f0c3bd94e432e650fad556a0/lib/Transforms/ConvertToCiphertextSemantics/ConvertToCiphertextSemantics.cpp#L535) per https://github.com/google/heir/pull/2347#discussion_r2464444433 

  Implements a DAG-based cost model for kernel transformations in the layout optimization pass, addressing issue #2316. Previously, `costOfKernelChange()` returned 0 for all kernels,
  leading to suboptimal layout conversion decisions. This PR introduces symbolic execution with the Halevi-Shoup baby-step giant-step algorithm to calculate exact rotation costs with
  O(√n) complexity.

  ## Motivation

  The layout optimizer needed accurate kernel costs to make informed decisions about when to hoist layout conversions. Without proper cost modeling, the optimizer couldn't effectively
  compare different layout+kernel combinations for operations like matrix-vector multiplication.

  ## Solution Approach

  Uses symbolic execution of kernel implementations to calculate exact rotation costs:

  1. **Creates symbolic inputs** - `SymbolicValue` with shape only, no data
  2. **Builds ArithmeticDAG** - Calls `implementHaleviShoup` for baby-step giant-step kernels
  3. **Counts rotations** - `RotationCountVisitor` traverses DAG with CSE
  4. **Returns cost** - Exact rotation count based on actual kernel structure

  ## Why This Approach?

  - ✅ **Exact counts** - Based on actual kernel DAG, not heuristics
  - ✅ **Efficient** - O(√n) rotations via Halevi-Shoup algorithm instead of O(n)
  - ✅ **Maintainable** - Reuses kernel generation code
  - ✅ **Auto-sync** - Kernel optimizations automatically reflected in costs
  - ✅ **Consistent** - Same rotation-counting as `computeCostOfLayoutConversion`

  ## Key Implementation Details

  ### Core Components

  - `costOfKernelChange()` - Main entry point in `LayoutOptimization.cpp`
  - `computeKernelCostFromDAG()` - Symbolic DAG builder with `implementHaleviShoup`
  - `RotationCountVisitor` - CSE-aware rotation counter in `lib/Kernel/RotationCountVisitor.h`
  - `getOperationShape()` - Tensor shape extractor

  ### Supported Kernels

  - **MatvecDiagonal** - Halevi-Shoup baby-step giant-step (O(√n) rotations)
  - **VecmatDiagonal** - Same algorithm as MatvecDiagonal
  - **Trivial** - Zero cost (elementwise operations)
 
  ## Example Impact

  For MatvecDiagonal on a 100×100 matrix using Halevi-Shoup:

  - **Before**: cost = 0 (ignored in optimization)
  - **After**: cost = 19 rotations (accurate O(√n) baby-step giant-step cost)

  The optimizer now makes informed decisions about layout hoisting based on actual kernel execution costs.

  ## Future Work

  - Issue #2351: Enhance cost model to include multiplicative depth
  - Issue #2352: Add integration test with multiple kernel alternatives

  ## Checklist

  - ✅ Signed Google CLA
  - ✅ Code follows Google C++ style
  - ✅ All tests pass
  - ✅ Unit tests cover edge cases
  - ✅ Documentation provided

  ## Related Issues

  Closes #2316

### CODE REVIEW

**Comment by j2kun** [comment]
Hey there! Thanks for the contribution. I will be able to give this PR a thorough review this week. In the meantime, could you run `pre-commit` on this PR as per https://heir.dev/docs/development/#pre-commit

**Comment by bon-cdp** [comment]
No problem. Sounds good.

File: docs/content/en/docs/Design/layout.md [hunk 1]
```diff
@@ -314,11 +314,12 @@ an estimate of the latency of a layout conversion, as well as the knowledge that
...
 IR.
 
-> **NOTE:** The cost of a kernel is currently considered free. This is mainly
-> because we don't have many different kernels for each op right now, so the
-> choice of kernel is not very interesting.
->
-> TODO(#2316): implement a cost model for kernels
+> **NOTE:** The cost of a kernel is calculated using symbolic execution
+> of kernel DAGs, as described in the kernel cost model implementation (see
+> `lib/Transforms/LayoutOptimization/LayoutOptimization.cpp:390-554`). The
```
Comment (j2kun) [inline docs/content/en/docs/Design/layout.md]: References to source code lines become obsolete very quickly. Let's just describe the method without referencing specific lines. If the implementation is in a separate file (maybe KernelCost.h/cpp) then it could make sense to point to the filepath.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 2]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
   return totalCost;
 }
 
+// Visitor to count UNIQUE rotations in a kernel DAG with CSE deduplication
+class KernelRotationCountVisitor {
+ public:
+  KernelRotationCountVisitor() {}
+
+  int64_t process(
+      const std::shared_ptr<kernel::ArithmeticDagNode<kernel::SymbolicValue>>&
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: For readability it may make sense to declare a (class scoped) using declaration like

`using NodeTy = kernel::ArithmeticDagNode<kernel::SymbolicValue>`

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 3]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+          node) {
+    const auto* nodePtr = node.get();
+
+    // If we've already visited this node, don't count it again (CSE)
+    if (visitedNodes.count(nodePtr)) {
+      return 0;
+    }
+    visitedNodes.insert(nodePtr);
+
+    return std::visit(
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Readability nit: separate these branches into separate functions and use type dispatching. See https://github.com/google/heir/blob/7aa0fe8285e0df05f0c3bd94e432e650fad556a0/lib/Kernel/IRMaterializingVisitor.h#L56 for an example.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 4]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+    case KernelName::MatvecNaive: {
+      if (shape.size() < 2) return 0;
+      kernel::SymbolicValue symbolicMatrix({shape[0], shape[1]});
+      kernel::SymbolicValue symbolicVector({shape[1]});
+      // MatvecNaive not yet implemented in KernelImplementation.h
+      // For now, use diagonal as approximation
+      kernelDag = kernel::implementMatvec(KernelName::MatvecDiagonal,
+                                          symbolicMatrix, symbolicVector);
+      break;
+    }
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Better is to return a `FailureOr<Cost>` and fail here, let the caller decide how to handle it (perhaps, return a very large sentinel cost or abort).

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 5]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+
+    case KernelName::MatmulDiagonal: {
+      if (shape.size() < 2) return 0;
+      // MatmulDiagonal typically implemented via multiple matvecs
+      // Conservative estimate: cost scales with both dimensions
+      return shape[0] + shape[1];
+    }
+
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Let's leave out matmul here since we don't have a kernel for it. I'm surprise it even has a `KernelName`; that is probably my oversight. Perhaps add a line pointing to https://github.com/google/heir/issues/1376 instead, and handle similarly to MatvecNaive

```
// TODO(#1376): evaluate bicyclic matmul kernel cost
```

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 6]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+  // Trivial kernels are free (elementwise operations)
+  if (newKernel == KernelName::Trivial) {
+    return 0;
+  }
+
+  // Extract operation dimensions
+  auto shape = getOperationShape(op);
+  if (!shape.has_value()) {
+    LLVM_DEBUG(llvm::dbgs() << "Could not extract shape for kernel cost\n");
+    return 0;  // Conservative: unknown ops cost nothing
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Similar to above, a better option would be to return a `FailureOr<Cost>` here, propagate any failures encountered, and have the caller explicitly skip the kernel option on failure instead of hoping the default costs are handled sensibly. In this case, a cost of zero would be incorrectly interpreted as "very cheap" rather than "conservative" as the comment states.

File: BUILD_AND_DEPLOY.md
Comment (j2kun) [inline BUILD_AND_DEPLOY.md:1]: Please revert this file, it is not necessary and the filename doesn't make sense. (Seems LLM-generated? If so please spot check your LLM outputs before asking people to review them)

File: MODULE.bazel.lock
Comment (j2kun) [inline MODULE.bazel.lock:1]: Please revert this file. We try to keep lockfile updates separate and created by automation.

File: tests/Transforms/layout_optimization/kernel_cost.mlir
Comment (j2kun) [inline tests/Transforms/layout_optimization/kernel_cost.mlir:1]: What's a bit strange about this PR is that there are no alternative kernels to choose from, so this test should not actually be verifying any new behavior. I think for now the unit tests are enough, and it would be worthwhile to file a GH issue for the task of adding such a test once there is a decision for which the cost of the kernel is a deciding factor.

File: lib/Kernel/KernelCostTest.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+#include "lib/Kernel/KernelImplementation.h"
+#include "lib/Kernel/KernelName.h"
+
+namespace mlir {
+namespace heir {
+namespace kernel {
+namespace {
+
+// Visitor to count UNIQUE rotations in a symbolic DAG with CSE deduplication
+class RotationCountVisitor {
```
Comment (j2kun) [inline lib/Kernel/KernelCostTest.cpp]: Is there a reason to have a separate copy of this Visitor from the one in LayoutOptimization? Maybe refactor it into a single class in `lib/Kernel/RotationCountVisitor.h/cpp`

File: lib/Kernel/KernelCostTest.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+  // Diagonal algorithm: one rotation per row
+  EXPECT_EQ(rotations, 4);
+}
+
+TEST(KernelCostTest, MatvecDiagonal_100x100_CostIs100) {
+  SymbolicValue matrix({100, 100});
+  SymbolicValue vector({100});
+
+  auto dag = implementMatvec(KernelName::MatvecDiagonal, matrix, vector);
```
Comment (j2kun) [inline lib/Kernel/KernelCostTest.cpp]: `implementMatvec` is the wrong kernel to use here, and this test demonstrates why. The actual kernel used is [`implementHaleviShoup`](https://github.com/google/heir/blob/7aa0fe8285e0df05f0c3bd94e432e650fad556a0/lib/Transforms/ConvertToCiphertextSemantics/ConvertToCiphertextSemantics.cpp#L535) and it uses closer to `sqrt(n)` rotations rather than `n` via an FHE technique called "baby step giant step." So the assertion should expect (if I recall from memory correctly) `2 sqrt(100) - 2 = 18` rotations. But you may need to double check.

I'm not even sure why we have `implementMatvec`, as it is only invoked in tests. Maybe it was just as a demonstration of the original `ArithmeticDag` framework before we had a baby-step giant-step implementation. It may be worth deleting in this PR.

File: lib/Kernel/KernelCostTest.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+  RotationCountVisitor counter;
+  int64_t rotations = counter.process(dag);
+
+  EXPECT_EQ(rotations, 512);
+}
+
+// Test that symbolic execution doesn't compute actual values
+TEST(KernelCostTest, SymbolicValueHasShapeOnly) {
```
Comment (j2kun) [inline lib/Kernel/KernelCostTest.cpp]: This test seems like it does nothing. Please delete.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 10]
```diff
@@ -385,10 +390,186 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+    default:
+      return 0;
+  }
+
+  if (!kernelDag) {
+    return 0;
+  }
+
+  // Count rotations in the symbolic DAG with CSE deduplication
+  KernelRotationCountVisitor counter;
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: This would be a good place to add a `// TODO(#issue_num)` to link this to an issue describing further enhancements of the cost model. E.g., including cost attributed to multiplicative depth, etc.

File: docs/content/en/docs/Design/layout.md (hunk 1 above)
Comment (bon-cdp) [reply docs/content/en/docs/Design/layout.md]: I modified documentation to reference 'lib/Kernel/RotationCountVisitor.h' instead of mentioning line numbers.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 2 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Added 'using NodeTy' in the 'RotationCountVisitor' class for readability.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 3 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Created separate 'operator()' methods for each node type, following this pattern.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 4 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: good call, computeKernelCostFromDAG now returns FailureOr<Cost>, so failure() will be returned for unsupported kernels.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 5 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: No problem. Done and comment added linking to issue.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 6 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Ok, in [c3f8b14] I tried to implement FailureOr<Cost> throughout. When shape extraction fails, I return a large sentinel cost (std::numeric_limits<Cost>::max() / 2) to properly discourage that kernel option instead of misleading with 0.

File: BUILD_AND_DEPLOY.md
Comment (bon-cdp) [reply BUILD_AND_DEPLOY.md:1]: My apologies, that wasn't supposed to be included, I'll try my best to ensure it doesn't happen again to not waste anyone's time.  md file removed.

File: MODULE.bazel.lock
Comment (bon-cdp) [reply MODULE.bazel.lock:1]: should be the same as the main branch again now.

File: tests/Transforms/layout_optimization/kernel_cost.mlir
Comment (bon-cdp) [reply tests/Transforms/layout_optimization/kernel_cost.mlir:1]: Deleted from tests/Transforms/layout_optimization/kernel_cost.mlir - will note to file the separate issue for adding a proper integration test for when we have multiple kernel implementations to choose from.

File: lib/Kernel/KernelCostTest.cpp (hunk 7 above)
Comment (bon-cdp) [reply lib/Kernel/KernelCostTest.cpp]: Good catch, refactored - lib/Kernel/RotationCountVisitor.{h,cpp} created as a shared lib with KernelCostTest.cpp and LayoutOptimization.cpp using the same implementation. Removed all duplicate code.

File: lib/Kernel/KernelCostTest.cpp (hunk 8 above)
Comment (bon-cdp) [reply lib/Kernel/KernelCostTest.cpp]: Nice, should be much more optimal now. Good reading and am currently taking baby steps to understand Halevi Shoup. 

The tested rotation counts ended up being higher than exactly sqrt(n) because I included Baby-step rotations at sqrt(n), giant-step rotations at another sqrt(n), and then any plaintext rotations for the Halevi-Shoup masks and post-processing for rectangular matrices. 

That said, the growth rate from the tests is ~O(sqrt(n))

File: lib/Kernel/KernelCostTest.cpp (hunk 9 above)
Comment (bon-cdp) [reply lib/Kernel/KernelCostTest.cpp]: deleted.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp (hunk 10 above)
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: Sounds good, will be cool to get there down the road.  See https://github.com/google/heir/issues/2351 for the issue.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 11]
```diff
@@ -385,10 +390,134 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+
+  switch (kernel) {
+    case KernelName::Trivial:
+      return 0;
+
+    case KernelName::MatvecDiagonal: {
+      if (shape.size() < 2) return failure();
+
+      // Use Halevi-Shoup baby-step giant-step algorithm for diagonal matvec
+      // This algorithm achieves O(sqrt(n)) rotations instead of O(n)
```
[the remark this answers is not in the document]
Comment (j2kun) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp:437]: This is a bad suggestion

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 12]
```diff
@@ -385,10 +390,134 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+  auto shape = getOperationShape(op);
+  if (!shape.has_value()) {
+    LLVM_DEBUG(llvm::dbgs() << "Could not extract shape for kernel cost\n");
+    // If we can't extract shape, skip this kernel option by returning
+    // a very large cost to discourage its selection
+    return std::numeric_limits<Cost>::max() / 2;
+  }
+
+  // Build symbolic DAG and count rotations
+  // TODO(#issue): enhance cost model to include multiplicative depth
```
[the remark this answers is not in the document]
Comment (j2kun) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: yes please make an actual issue for this or attach it to an existing issue. We use https://github.com/j2kun/todo-backlinks to automatically annotate issues with links to the corresponding TODOs in the codebase.

[the remark this answers is not in the document]
Comment (bon-cdp) [reply lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]: I created https://github.com/google/heir/issues/2351 for this (not being able to find any issues it fit nicely in) and updated the comment in the code.

File: tests/Transforms/layout_optimization/kernel_cost.mlir
Comment (bon-cdp) [reply tests/Transforms/layout_optimization/kernel_cost.mlir:1]: I created the first draft of https://github.com/google/heir/issues/2352 for this.

File: lib/Transforms/LayoutOptimization/LayoutOptimization.cpp [hunk 13]
```diff
@@ -385,10 +390,134 @@ Cost LayoutOptimization::costOfChangedResult(Operation* kernel,
...
+          symbolicVector, symbolicMatrix, originalShape);
+
+      if (!kernelDag) return failure();
+
+      RotationCountVisitor counter;
+      return counter.process(kernelDag);
+    }
+
+    case KernelName::MatvecNaive:
+      // TODO: implement MatvecNaive kernel cost model
```
Comment (j2kun) [inline lib/Transforms/LayoutOptimization/LayoutOptimization.cpp]:
```suggestion
      // TODO(#1589): implement MatvecNaive kernel cost model
```

File: lib/Kernel/KernelCostTest.cpp [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+  SymbolicValue matrix({100, 100});  // 100 diagonals in diagonal packing
+  std::vector<int64_t> originalShape = {100, 100};
+
+  auto dag = implementHaleviShoup(vector, matrix, originalShape);
+
+  RotationCountVisitor counter;
+  int64_t rotations = counter.process(dag);
+
+  // Halevi-Shoup for 100x100: includes baby-step, giant-step, and plaintext
+  // rotations Actual measured: 119 rotations
```
Comment (j2kun) [inline lib/Kernel/KernelCostTest.cpp]: 119 rotations for a 100x100 matrix is far too big (it's not close to sqrt(n)), and I think the problem is that _plaintext_ rotations should not be counted. Those should be considered free in this cost model (they happen in cleartext!).

This should not be hard to fix, but the use of `SymbolicValue` in the arithmetic dag will not suffice alone, since it does not carry information about what is secret. The simplest path forward is likely to add an extra `isSecret` boolean field to `SymbolicValue`, and make it false for the matrix input and true for the vector input, and then the visitor will need to track the derived `isSecret` at each Node, which can be achieved by having the visitor internally store a `map` from the `Node` to the `isSecret` boolean, and inserting into it as it visits the DAG.

Comment (bon-cdp) [reply lib/Kernel/KernelCostTest.cpp]: My apologies, I wasn't sure if there would be some use for counting plaintext rotations into the future (although seeing as they're free anyways this seems kinda silly in retrospect). The 119 rotation result did throw me off in any regard, so I'll be sure to note concerns of said nature more plainly when they arise going forward. 

Here I tried to implement the solution as you suggested:

  1. Added isSecret field to SymbolicValue to track whether values are encrypted
  2. Added nodeSecretStatus map to visitor to derive and store secret status for each DAG node during traversal
  3. Visitor derives and tracks secret status as it traverses:
    - LeafNode: Gets from node.value.isSecret()
    - Constants: Always false (plaintext)
    - Binary ops: leftIsSecret || rightIsSecret
    - LeftRotateNode: Only counts rotation if operand is secret
  4. Updated call sites: Vector=true (encrypted), Matrix=false (plaintext)

100x100 matrix is now 19 rotations
512x512 matrix is now 47 rotations. 

p.s. Sorry for the multiple identical commits from [11440e8](https://github.com/google/heir/pull/2347/commits/11440e8ed57e448fbb147e9276265c2f703be77a) to [3ef6b90](https://github.com/google/heir/pull/2347/commits/3ef6b9085e172d37bf1ef319989690284aa82c6a) , the cla wasnt passing with your account as a coauthor for what I auto committed here [8ab4c03](https://github.com/google/heir/pull/2347/commits/8ab4c03b0b1110434daddaf662f3fd12aa2b3090)

Comment (j2kun) [reply lib/Kernel/KernelCostTest.cpp]: > 100x100 matrix is now 19 rotations
> 512x512 matrix is now 47 rotations.

excellent!

File: lib/Kernel/RotationCountVisitor.h [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace mlir {
+namespace heir {
+namespace kernel {
+
+// Visitor to count unique rotations in a symbolic DAG with CSE deduplication.
+// This visitor traverses an ArithmeticDAG and counts rotation operations,
+// ensuring that shared subexpressions (common subexpression elimination) are
+// only counted once.
+class RotationCountVisitor {
```
Comment (j2kun) [inline lib/Kernel/RotationCountVisitor.h:20]: Unless you tried this already and it didn't work for some reason, the CSE/caching behavior is already handled by an existing base class `CachingVisitor` (https://github.com/google/heir/blob/2fa61cc0858d0ff48e9d876965acdf3c085da97a/lib/Kernel/ArithmeticDag.h#L208). This would likely simplify the implementation of this visitor a bit.

**Review by j2kun — APPROVED** [review]
One last suggestion, and then LGTM!

When you're ready, please squash all your commits into one (a weird quirk of the google-internal part of the merging process) rebase over main, and apply pre-commit formatting.

Thanks so much for your contribution! Let me know if you want to brainstorm further work on this or any other HEIR topics.

File: lib/Kernel/RotationCountVisitor.h (hunk 15 above)
Comment (bon-cdp) [reply lib/Kernel/RotationCountVisitor.h:20]: I tried implementing this and it failed 8/9 tests (the 2x2 small matrix test passed). It ends up scaling with n and not sqrt(n), along with exponential blowup at large sizes (which suggests there's also multiplication happening - not just linear reuse, but potentially nested shared subexpressions being counted multiple times at each level).

I think the issue is how CachingVisitor handles repeat visits vs what we need for CSE.

**My recursive CSE deduplication:**
```
lib/Kernel/RotationCountVisitor.cpp:20-28
  int64_t RotationCountVisitor::processInternal(
      const std::shared_ptr<NodeTy>& node) {
    const auto* nodePtr = node.get();

    if (visitedNodes.count(nodePtr)) {
      return 0;  // already visited this node
    }
    visitedNodes.insert(nodePtr);
    // ...
  }
```

  1. Process Add → visit left Rot
  2. First time seeing Rot: not in visitedNodes
  3. Process Rot, mark as visited → returns 1
  4. Visit right Rot (same node)
  5. Already in visitedNodes → return 0
  6. Add: 1 + 0 = 1 rotation 


**CachingVisitor Approach:**
 ```
 lib/Kernel/ArithmeticDag.h:213-224
  ResultType process(const std::shared_ptr<ArithmeticDagNode<T>>& node) {
    const auto* nodePtr = node.get();
    if (auto it = cache.find(nodePtr); it != cache.end()) {
      return it->second;  // Returns cached RotationResult
    }
    ResultType result = std::visit(*this, node->node_variant);
    cache[nodePtr] = result;
    return result;
  }
```

  Execution with RotationResult{count, isSecret}:
  1. Process Add → visit left Rot
  2. First time: compute Rot → {1, true}, cache it
  3. Visit right Rot (same node)
  4. Found in cache → return {1, true}
  5. Add: {1, true} + {1, true} = {2, true} 

**Why Exponential Blowup Happens with CachingVisitor:**

With nested shared subexpressions in Halevi-Shoup:

             Add
            /   \
          Mul   Mul
          / \   / \
        R1  R2 R2  R3  (R2 shared between Mul nodes)
       /    |   |   \
    (subtree with 10 rotations)

  CachingVisitor:
  - R2 subtree computed once: {10, true}
  - Mul(left): R1{10} + R2{10} = {20, true}
  - Mul(right): R2{10 again!} + R3{10} = {20, true}
  - Add: {20} + {20} = {40, true}  Should be ~30!

 Every level of sharing multiplies the count. With deep DAGs, this becomes exponential.

**Semantic Difference:**
CachingVisitor solves: "What is the local property of this subtree?"
  - Memoizes computation results
  - If subtree is reused, return the same answer
  - Used for: evaluation, type inference, depth calculation

  My implementation solves: "What is the global count of unique nodes?"
  - Tracks "have I seen this node anywhere in the entire DAG?"
  - If node is reused, don't count it again (return 0)
  - Used for: CSE deduplication, unique operation counting

**TLDR:**
IMHO without proper CSE deduplication, each rotation node gets counted every time it's visited instead of once. The Halevi-Shoup algorithm creates a DAG where rotation nodes are reused many times, so the count goes to O(n) instead of O(sqrt(n)). I could also just be using it wrong. 

Let me know what you think.

**Comment by bon-cdp** [comment]
Thank you for all the great advice and support through the process!

 I learned a ton, and would love to get involved in any way I can.

I'd be interested in hearing your thoughts on https://github.com/google/heir/issues/2351 as it seems like a natural next step - or if you have something else in mind I'm all ears. I'll look into [update research_with_heir with more pubs](https://github.com/google/heir/commit/a702be88df1e497eb0a2c211b309163ba2beeb86) a bit more too. https://github.com/google/heir/issues/2318 for & https://github.com/google/heir/issues/2364 also look like a fun challenges. 

Tbh I'm super dumb with math (just starting to learn) so even menial things you think would improve my overall understanding I'd be happy to do. 

Feel free to mention/assign me in any other issues I could look at as well.

File: lib/Kernel/RotationCountVisitor.h (hunk 15 above)
Comment (j2kun) [reply lib/Kernel/RotationCountVisitor.h:20]: That makes sense. I think after this PR is merged I will refactor into a new base class to handle this in general, since I think it will be useful elsewhere.

Comment (bon-cdp) [reply lib/Kernel/RotationCountVisitor.h:20]: Sounds good.

**Review by j2kun — APPROVED** [review]
